L1Trigger/Phase2L1GMT/src/KMTF.cc

```diff
@@ -0,0 +1,482 @@
+    if (stub->depthRegion() == 1) {
+      if (Nstubs1 < MAXN) {
+        stubs1.push_back(stub);
+        Nstubs1++;
+      }
+    }
+  }
+
+  //Sort the seeds by tag so that the emulator is aligned like the firmware
+  SeedSorter sorter;
+  if (stubs4.size() > 1) {
+    std::sort(stubs4.begin(), stubs4.end(), sorter);
+  }
+  if (stubs3.size() > 1) {
+    std::sort(stubs3.begin(), stubs3.end(), sorter);
+  }
+  if (stubs2.size() > 1) {
+    std::sort(stubs2.begin(), stubs2.end(), sorter);
+  }
+  if (stubs1.size() > 1) {
+    std::sort(stubs1.begin(), stubs1.end(), sorter);
+  }
+
+  bool pre_patterns = (verbose_ > 1) && ((Nstubs4 + Nstubs3 + Nstubs2 + Nstubs1) > 2);
+
+  //OK now process the data almost as in hardware
+  for (unsigned int i = 0; i < 32; ++i) {
+    //print the stubs taking into account
+
+    bool patterns = pre_patterns && ((i < Nstubs4) || (i < Nstubs3) || (i << Nstubs2));
+
+    if (patterns) {
+      printf("KMTFPattern ");
+      if (i < Nstubs4)
+        printf("%d %d %d 1 %d 0 ",
+               stubs4[0]->coord1(),
+               stubs4[0]->coord2(),
+               stubs4[0]->quality(),
+               stubs4[0]->kmtf_address());
+      else
+        printf("0 0 0 0 511 0 ");
+
+      if (i < Nstubs3) {
+        for (const auto& s : stubs3) {
+          printf("%d %d %d 1 %d 0 ", s->coord1(), s->coord2(), s->quality(), s->kmtf_address());
+        }
+        //pad with zeros
+        for (unsigned int j = stubs3.size(); j < 32; ++j) {
+          printf("0 0 0 0 511 0 ");
+        }
+      } else {
+        for (unsigned int j = stubs3.size(); j < 32; ++j) {
+          printf("0 0 0 0 511 0 ");
+        }
+        for (const auto& s : stubs3) {
+          printf("%d %d %d 1 %d 0 ", s->coord1(), s->coord2(), s->quality(), s->kmtf_address());
+        }
+      }
+
+      if (i < Nstubs2) {
+        for (const auto& s : stubs2) {
+          printf("%d %d %d 1 %d 0 ", s->coord1(), s->coord2(), s->quality(), s->kmtf_address());
+        }
+        //pad with zeros
+        for (unsigned int j = stubs2.size(); j < 32; ++j) {
+          printf("0 0 0 0 511 0 ");
+        }
+      } else {
+        for (unsigned int j = stubs2.size(); j < 32; ++j) {
+          printf("0 0 0 0 511 0 ");
+        }
+        for (const auto& s : stubs2) {
+          printf("%d %d %d 1 %d 0 ", s->coord1(), s->coord2(), s->quality(), s->kmtf_address());
+        }
+      }
+      if (i < Nstubs1) {
+        for (const auto& s : stubs1) {
+          printf("%d %d %d 1 %d 0 ", s->coord1(), s->coord2(), s->quality(), s->kmtf_address());
+        }
+        //pad with zeros
+        for (unsigned int j = stubs1.size(); j < 32; ++j) {
+          printf("0 0 0 0 511 0 ");
+        }
+      } else {
+        for (unsigned int j = stubs1.size(); j < 32; ++j) {
+          printf("0 0 0 0 511 0 ");
+        }
+        for (const auto& s : stubs1) {
+          printf("%d %d %d 1 %d 0 ", s->coord1(), s->coord2(), s->quality(), s->kmtf_address());
+        }
```

Comment:
These printf() are triggered by the verbose level in the config( off by default). Since these are used in the firmware/emulator agreement test, we prefer to keep them here.